File: 01/python/solution.py
def has_a_sum(arr, K):

	# Return false if array has 1 or less elements
	if(len(arr) < 2):
		return False

	# Init hash set
	cache = set()

	# One pass over array
	for a in arr:

		# If difference is already in cache return true
		if(a in cache):
			return True
		# Otherwise add differnce to cache
		else:
			cache.add(K - a)

	# No sum was found
	return False

if __name__ == '__main__':
	
	# Read test file
	f = open("../testcase.txt", "r")

	# Number of test cases
	T = int(f.readline())

	# Loop over test cases
	for __ in xrange(T):

		# Read length of array N and sum K
		N,K = map(int, f.readline().split())

		# Fill array
		arr = map(int, f.readline().split())

		# Print solution
		print("Case #%d: %s" % (__+1, has_a_sum(arr, K)))

	# Close file
	f.close()

File: 01/cpp/solution.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <set>

using namespace std;

bool has_a_sum(const std::vector<int> & arr, const int k){
    
    // Return false if array has 1 or less elements
    if(arr.size() < 2)
        return false;

    // Init hash set
    set<int> cache;

    // One pass over array
    for(int i = 0; i < arr.size(); i++){

        // If difference is already in cache return true
        if(cache.find(arr[i]) != cache.end()){
            return true;
        }
        // Otherwise add differnce to cache
        else{
            cache.insert(k - arr[i]);
        }
    }

    // No sum was found
    return false;
}

int main(){

    // Read test file
    ifstream inFile;
    inFile.open("../testcase.txt");

    if(!inFile){
        cerr << "Unable to open the file" << endl;
        return -1;
    }

    // Number of test cases
    int T;
    inFile >> T;

    // Loop over test cases
    for(int t = 1; t <= T; t++){

        // Read length of array N and sum K
        int N,K;
        inFile >> N >> K;

        // Fill array
        vector<int> arr(N);
        for(int n = 0; n < N; n++){
            inFile >> arr[n];
        }

        // Print solution
        cout << "Case #" << t << ": " << has_a_sum(arr, K) << endl;
    }

    // Close file
    inFile.close();

    return 0;
}